### PR TITLE
Adding support to specify "level" parameter  for `Dataframe.rename` 

### PR DESCRIPTION
**Pull request by skirui-source**

### CODE REVIEW

**Comment by karthikeyann** [comment]
Converting to draft. (in progress), removing [WIP].

**Comment by kkraus14** [comment]
@skirui-source it looks like a bunch of additional commits are pulled into this PR for some reason and it makes it really hard to review the actual changes. I'm guessing it needs to have the upstream `branch-0.18` merged into it. Could you give it a shot?

File: python/cudf/cudf/core/dataframe.py
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:3430]: Because we're reaching directly into `_source_data` here we need to be careful to handle the cases where there's duplicate named columns so the names in `self.index.names` doesn't match the names in `self.index._source_data`.

This also looks like it only handles a level name while it should also allow an integer representing the index of the MultiIndex level.

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/column_accessor.py
Comment (isVoid) [inline python/cudf/cudf/core/column_accessor.py:344]: We should use underscore in front of the internal function names, assuming it is one.

File: python/cudf/cudf/core/column_accessor.py
Comment (isVoid) [inline python/cudf/cudf/core/column_accessor.py]: We should be cautious not to assign to `self` as this will throw the type checker off. We should be able to comfortably get rid of this line here.
```suggestion
            ca = ColumnAccessor(dict(zip(new_names, columns)), level_names=self.level_names, multiindex=self.multiindex)
```

File: python/cudf/cudf/core/column_accessor.py
Comment (isVoid) [inline python/cudf/cudf/core/column_accessor.py]: Similarly
```suggestion
                ca = ColumnAccessor(dict(zip(new_names, columns)), level_names=self.level_names, multiindex=self.multiindex)
        
        return ca
```

File: python/cudf/cudf/core/column_accessor.py
Comment (isVoid) [inline python/cudf/cudf/core/column_accessor.py:362]: Let's be cautious not to asisgn to `self` as this may throw the type checker off. We should be able to comfortably remove this line here.
```suggestion
            ca = ColumnAccessor(
                dict(zip(new_names, columns)),
                level_names=self.level_names,
                multiindex=self.multiindex,
            )
```

File: python/cudf/cudf/core/column_accessor.py
Comment (isVoid) [inline python/cudf/cudf/core/column_accessor.py:381]: Similarly
```suggestion
                ca = ColumnAccessor(
                    dict(zip(new_names, columns)),
                    level_names=self.level_names,
                    multiindex=self.multiindex,
                )

        return ca
```

File: python/cudf/cudf/core/dataframe.py
Comment (isVoid) [inline python/cudf/cudf/core/dataframe.py:3497]: Do we still accept `columns` as a function? As indicated in docstrings.

**Review by isVoid — COMMENTED** [review]
Some minor issues, plus an extra file commited that should be removed.

**Comment by isVoid** [comment]
Apparently the review system devoured my previous comment then spitted it out again. Removed redundant comments.

**Review by karthikeyann — CHANGES_REQUESTED** [review]
an artifact of merge still left.
remove binary file `python/cudf/cudf/core/Cisco Webex Meetings.lnk`

File: python/cudf/cudf/tests/test_dataframe.py
Comment (skirui-source) [reply python/cudf/cudf/tests/test_dataframe.py:8364]: but for this case, i am testing level support for regular (non-multicolumns) so i don't understand why i should test rename on index parameter?  please clarify

Comment (isVoid) [reply python/cudf/cudf/tests/test_dataframe.py:8364]: I take that as a corner case where dataframe with regular index should also support specifying `level=0`, such as
```
>>> pdf = pd.DataFrame({'a': [1,2,3], 'b':[4,5,6], 'c':[7,8,9]})
>>> pdf.rename(index={1:123}, level=0)
     a  b  c
0    1  4  7
123  2  5  8
2    3  6  9
```
Especially in this case, I'm interested because the index is a `RangeIndex`, renaming it would require instantiating it into a `Int64Index`.

File: python/cudf/cudf/core/dataframe.py
Comment (skirui-source) [reply python/cudf/cudf/core/dataframe.py:3497]: I just confirmed with Ashwin this PR will only implement` index ` and  `columns` as Mapper  only

Comment (isVoid) [reply python/cudf/cudf/core/dataframe.py:3497]: Should we label this PR as `breaking`? Since we used to support function and now we don't. @shwina

**Review by isVoid — APPROVED** [review]
lgtm, should we label the PR as `breaking`? Since it used to support `columns` as a function and now we don't. @shwina

File: python/cudf/cudf/core/dataframe.py
Comment (isVoid) [inline python/cudf/cudf/core/dataframe.py:3507]: I think originally there is a case handling `columns` as functions, so I think it should be considered in the new implementation.

**Review by isVoid — CHANGES_REQUESTED** [review]

**Comment by harrism** [comment]
Should this be moved to 0.19?

File: python/cudf/cudf/core/dataframe.py
Comment (skirui-source) [reply python/cudf/cudf/core/dataframe.py:3507]: I just pushed the changes to fix this. Please review and let me know .

**Review by isVoid — APPROVED** [review]
lgtm!

File: python/cudf/cudf/core/column_accessor.py
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:345]: For the following to work, `Mapping`, `Any` and `Callable` must be imported from the `typing` module:

```suggestion
        self, mapping: [Mapping[Any, Any], Callable], level: int = None
```

File: python/cudf/cudf/core/column_accessor.py
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:349]: I think we could expand this docstring a bit. In particular, we should explain the `mapping` parameter in more detail: i.e., what the behaviour is when `mapping` is a dictionary v/s a callable.

File: python/cudf/cudf/core/column_accessor.py
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:346]:
```suggestion
    ) -> ColumnAccessor:
```

Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py:346]: When adding type annotations, also add the return type.

File: python/cudf/cudf/core/column_accessor.py
Comment (isVoid) [reply python/cudf/cudf/core/column_accessor.py:345]: Should `level` be `Optional[int]` instead?

File: python/cudf/cudf/core/column_accessor.py
Comment (skirui-source) [reply python/cudf/cudf/core/column_accessor.py:346]: not sure what you mean by this, can you please clarify?

File: python/cudf/cudf/core/column_accessor.py
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:388]: The style check is failing here because `level` is optional, but here it is expected to be an integer. I think we should make `level` non-optional.

**Review by shwina — APPROVED** [review]
LGTM pending style issues. Really great work @skirui-source!

File: python/cudf/cudf/core/column_accessor.py
Comment (shwina) [inline python/cudf/cudf/core/column_accessor.py:352]:
```suggestion
        self, mapper: Union[Mapping[Any, Any], Callable], level: int
```

Let's make this a required (non-optional arg)

**Comment by mike-wendt** [comment]
Updated target branch to `branch-0.19` and project after talking to @kkraus14

**Review by karthikeyann — APPROVED** [review]
Approving for unblocking. (no cpp file changes)

**Comment by kkraus14** [comment]
Can you merge `branch-0.19` into this? Looks like it's showing some incorrect changes.

File: ci/gpu/build.sh
Comment (kkraus14) [inline ci/gpu/build.sh:86]: Thanks for fixing this 😄

File: ci/gpu/build.sh
Comment (kkraus14) [inline ci/gpu/build.sh:223]: Is this an intended change?

File: python/cudf/cudf/core/dataframe.py
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:3466]: This is a deep copy, do we need a deep copy here?

File: python/cudf/cudf/core/column_accessor.py
Comment (kkraus14) [inline python/cudf/cudf/core/column_accessor.py:383]: We should use a list comprehension here instead of initializing an empty list and then appending to it.

File: python/cudf/cudf/core/column_accessor.py
Comment (kkraus14) [inline python/cudf/cudf/core/column_accessor.py:394]: List comprehension here as well

File: python/cudf/cudf/core/column_accessor.py
Comment (kkraus14) [inline python/cudf/cudf/core/column_accessor.py:344]: Do we need to implement this all the way down at the `ColumnAccessor`? This feels like something we should be able to accomplish in `Frame` instead.

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/column_accessor.py
Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py:344]: It was my suggestion initially to move the logic here. But after giving it some thought I think I agree with Keith here.

ColumnAccessor was meant initially to be really a dict-like object but it has grown into something much more than that.

Likely we want to move this (and other methods in this class) into either free functions that operate on and return CAs, or directly into the Frame class.

Comment (kkraus14) [reply python/cudf/cudf/core/column_accessor.py:344]: Even if we make them free methods we should have them operate on Frames instead of ColumnAccessor objects whenever possible.

Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py:344]: That just makes those methods do too much, since a rename is exclusively an operation on the keys of a column accessor.

Comment (shwina) [reply python/cudf/cudf/core/column_accessor.py:344]: So my suggestion would be something like:

```python
def rename_levels(ca: ColumnAccessor, levels, mapper):
     """Rename the specified levels of the given ColumnAccessor"""
    pass

class Frame:
     def rename(self, ...):
          new_data = rename_levels(self._data, ...)
          return self.__class__._from_data(new_data)
```

I think the above separates concerns a little better than implementing _all_ the logic in `Frame.rename`. Conceivably, renaming the levels of a ColumnAccessor could be useful outside of `Frame.rename`.

I also think it's better to write functions that operate directly on ColumnAccessors rather than Frames, as there might be situations where we want to compose ColumnAccessor operations  (e.g., insert_level + rename).

File: ci/gpu/build.sh
Comment (shwina) [reply ci/gpu/build.sh:223]: This change already exists in branch-0.19 -- is GitHub just derping out here?

https://github.com/rapidsai/cudf/blob/branch-0.19/ci/gpu/build.sh#L227

File: python/cudf/cudf/core/column_accessor.py
Comment (kkraus14) [reply python/cudf/cudf/core/column_accessor.py:344]: Fair enough, lets move forward with it as a ColumnAccessor function

File: ci/gpu/build.sh
Comment (kkraus14) [reply ci/gpu/build.sh:223]: Doesn't look like it and this change needs to be reversed

Comment (skirui-source) [reply ci/gpu/build.sh]: @kkraus14 I reversed this change on Friday, awaiting your approval

**Comment by kkraus14** [comment]
@skirui-source thanks, changes look good now, but looks like you have some test failures that need to be resolved before we can merge. Can you take a look?

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by kkraus14** [comment]
Looks like there's a couple of failed tests that need to be addressed.

**Comment by shwina** [comment]
The failing test is this one:

https://github.com/rapidsai/cudf/blob/branch-0.19/python/cudf/cudf/tests/test_dataframe.py#L652-

It feels we had some custom, non-Pandas behaviour before this PR that we should consider dropping. Suffixing like this doesn't always work; for example on branch-0.18:

```python
In [4]: a = cudf.DataFrame({'a': [1, 2, 3], "b": [4, 5, 6], "z_1": [2, 3, 4]})

In [5]: a
Out[5]:
   a  b  z_1
0  1  4    2
1  2  5    3
2  3  6    4

In [6]: a.rename(columns={'a': 'z', 'b': 'z'})
Out[6]:
   z  z_1
0  1    2
1  2    3
2  3    4
```

**Comment by kkraus14** [comment]
Got it. We should probably throw in that situation and we should probably test that we throw in that situation?

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by kkraus14** [comment]
rerun tests

**Pull request merged**
